engines/cegar_values.cpp

```diff
@@ -0,0 +1,309 @@
+** in the top-level source directory) and their institutional affiliations.
+** All rights reserved.  See the file LICENSE in the top-level source
+** directory for licensing information.\endverbatim
+**
+** \brief A simple CEGAR loop that abstracts values with frozen variables
+**        and refines by constraining the variable to the value again
+**
+**/
+
+#include "engines/cegar_values.h"
+
+#include <unordered_set>
+
+#include "core/fts.h"
+#include "core/rts.h"
+#include "engines/ceg_prophecy_arrays.h"
+#include "engines/ic3ia.h"
+#include "smt-switch/identity_walker.h"
+#include "smt/available_solvers.h"
+#include "utils/exceptions.h"
+#include "utils/logger.h"
+#include "utils/make_provers.h"
+
+using namespace smt;
+using namespace std;
+
+namespace pono {
+
+static unordered_set<PrimOp> nl_ops({ Mult,
+                                      Div,
+                                      Mod,
+                                      Abs,
+                                      Pow,
+                                      IntDiv,
+                                      BVMul,
+                                      BVUdiv,
+                                      BVSdiv,
+                                      BVUrem,
+                                      BVSrem,
+                                      BVSmod });
+
+/** Class for abstracting values with frozen variables
+ */
+class ValueAbstractor : public smt::IdentityWalker
+{
+ public:
+  ValueAbstractor(TransitionSystem & ts, UnorderedTermMap & abstracted_values)
+      : smt::IdentityWalker(ts.solver(), false),
+        ts_(ts),
+        abstracted_values_(abstracted_values)
+  {
+  }
+
+ protected:
+  smt::WalkerStepResult visit_term(smt::Term & term) override
+  {
+    if (!preorder_) {
+      if (term->is_value() && term->get_sort()->get_sort_kind() != ARRAY) {
+        // create a frozen variable
+        Term frozen_var =
+            ts_.make_statevar("__abs_" + term->to_string(), term->get_sort());
+        // will be made frozen later
+        // since the transition system will be modified after this
+        save_in_cache(term, frozen_var);
+        abstracted_values_[frozen_var] = term;
+        return Walker_Continue;
+      }
+
+      Op op = term->get_op();
+      if (!op.is_null() && nl_ops.find(op.prim_op) == nl_ops.end()) {
+        // only rebuild terms with operators that can't
+        // create nonlinearities by replacing constant values
+        TermVec cached_children;
+        Term c;
+        for (const auto & t : term) {
+          c = t;
+          query_cache(t, c);
+          cached_children.push_back(c);
+        }
+        save_in_cache(term, solver_->make_term(op, cached_children));
+      } else {
+        save_in_cache(term, term);
+      }
+    }
+    return Walker_Continue;
+  }
+
+  TransitionSystem & ts_;
+  UnorderedTermMap & abstracted_values_;
+};
+
+TransitionSystem create_fresh_ts(bool functional, const SmtSolver & solver)
```

Comment:
how about putting this method in the ts.h as a static function?

Reply:
Great idea! It didn't work nicely to put in `ts.h` because it doesn't know about `Relational` or `Functional` `TransitionSystem`, but I created a new utility file in 6c7b9bc066942633f5447df8c9c8ce10ee1ce7da.